package_modules/apt_get

```diff
@@ -0,0 +1,259 @@
+    for line in sys.stdin:
+        result = re.match("^File=(.*)", line)
+        if (result):
+            pkg_string = result.group(1)
+            break
+
+    if (pkg_string == ""):
+        return 1
+
+    repo = False
+    if (re.match("^/.*", pkg_string)):
+        # Absolute file.
+        sys.stdout.write("PackageType=file\n")
+    elif (re.search("([:,]|-[0-9])", pkg_string)):
+        # Contains either a version number or an illegal symbol.
+        sys.stdout.write(line + "ErrorMessage: Package string with illegal format\n")
+        return 1
+    else:
+        repo = True
+        sys.stdout.write("PackageType=repo\n")
+
+    if (repo):
+        sys.stdout.write("Name=" + pkg_string + "\n")
+        return 0
+    else:
+        sys.stdout.flush()
+        return subprocess.call([dpkg_deb_cmd, "--showformat", dpkg_output_format, "-W", result.group(1)])
+
+
+def list_installed():
+    for line in sys.stdin:
+        # Ignore everything.
+        pass
+
+    process = subprocess.Popen([dpkg_query_cmd, "--showformat", dpkg_status_format, "-W"], stdout=subprocess.PIPE)
+    installed_package = False
+    for line in process.stdout:
+        if line[-1] == "\n":
+            line = line[:-1]
+        # The space before "installed" is important, because it can be "not-installed".
+        if re.match("^Status=.* installed.*", line):
+            installed_package = True
+        elif re.match("^Status=.*", line):
+            installed_package = False
+        elif installed_package:
+            sys.stdout.write(line + "\n")
+
+    return 0
+
+
+def list_updates(online):
+    for line in sys.stdin:
+        # Ignore everything.
+        pass
+
+    if online:
+        result = subprocess.call([apt_get_cmd] + apt_get_options + ["update"], stdout=sys.stderr)
+        if (result != 0):
+            return result
+
+    process = subprocess.Popen([apt_get_cmd] + apt_get_options + ["-s", "upgrade"], stdout=subprocess.PIPE)
+    for line in process.stdout:
+        # Example of lines that we try to match:
+        #        (name)  (old version (ignored))  (new version)     (repository(ies) (ignored))  (arch)
+        #           |              |                    |                        |                 |
+        #           V              V                    V                        V                 V
+        #   Inst php5-cli [5.3.10-1ubuntu3.17] (5.3.10-1ubuntu3.18 Ubuntu:12.04/precise-updates [amd64]) []
+        #
+        # Note architecture included in the name on this one:
+        #   Inst php5-cli:i386 [5.3.10-1ubuntu3.17] (5.3.10-1ubuntu3.18 Ubuntu:12.04/precise-updates [i386]) []
+        #
+        # Note multiple repositories in this one:
+        #   Inst linux-libc-dev [2.6.32-48squeeze4] (2.6.32-48squeeze6 Debian:6.0.10/oldstable, Debian-Security:6.0/oldoldstable [amd64])
+        #
+        match = re.match("^Inst\s+([^\s:]+)(?::\S+)?\s+\[[^]\s]+\]\s+\((\S+)\s+[^,\s]+(?:,\s+[^,\s]+)*\s+\[([^]\s]+)\]\).*", line)
```

Comment:
Now, quite apart from how hard this is to review (because I have to page down six times to find the scroll-bar I need in order to see the whole thing, then page back up again), this is illegible gibberish (as regexes so often are), which means that someone reading the code can't really tell if it does what your (excellent) comment claims.  It's wonderfully terse but anti-debuggable.
```
words = line.split()
if not words or words.pop(0) != 'Inst' or not words: continue
name = words.pop(0)
if words[0].startswith('['): words = words.pop(0) # discard old version, if present
if not words[0].startswith('('): continue
version = words.pop(0)[1:]
while not words[0].endswith(')'): words.pop(0) # discard repositories
arch = words.pop(0)[:-1]
if arch.startswith('[') and arch.endswith(']'): arch = arch[1:-1]
else: continue
sys.stdout.write('Name=%s\nVersion=%s\nArch=%s\n' % (name, version, arch))
```
I don't think the reader is in any doubt that this parses what you have above; it's possible to add comments where they belong; and someone looking for a bug can probably see where it is.